File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
    private class Node{
        int key;
        int value;
        Node next;
        Node prev;
    }

    private final int capacity;
    private final Map<Integer, Node> map;
    private final Node head;
    private final Node tail;
    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.map = new HashMap<>();
        this.head = new Node();
        this.tail = new Node();
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        Node node = map.get(key);
        if (node == null) {
            return -1;
        }
        moveToHead(node);
        return node.value;
    }

    private void moveToHead(Node node) {
        removeNode(node);
        addToHead(node);
    }

    private void removeNode(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void addToHead(Node node) {

        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }
    
    public void put(int key, int value) {
        Node node = map.get(key);
        if (node != null) {
            node.value = value;
            moveToHead(node);
        } else {
            node = new Node();
            node.key = key;
            node.value = value;

            if (map.size() == capacity) {
                map.remove(tail.prev.key);
                removeNode(tail.prev);
            }
            map.put(key, node);
            addToHead(node);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}
        self.capacity = capacity
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def addToFront(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node
        
    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self._remove(node)
            self.addToFront(node)
            
            return node.val
        
        return -1
        

    def put(self, key: int, value: int) -> None:
        if key not in self.cache:
            if len(self.cache) == self.capacity:
                nodeToRemove = self.tail.prev
                del self.cache[nodeToRemove.key]
                self._remove(nodeToRemove)
            newNode = Node(key, value)
            self.cache[key] = newNode
            self.addToFront(newNode)
        else:
            node = self.cache[key]
            node.val = value
            self._remove(node)
            self.addToFront(node)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

# class LRUCache:
#     def __init__(self, capacity: int):
#         self.capacity = capacity
#         self.cache = {}
#         self.head = Node()
#         self.tail = Node()
#         self.head.next = self.tail
#         self.tail.prev = self.head

#     def addToFront(self, node):
#         node.prev = self.head
#         node.next = self.head.next
#         self.head.next.prev = node
#         self.head.next = node

#     def removeNode(self, node):
#         node.prev.next = node.next
#         node.next.prev = node.prev

#     def get(self, key: int) -> int:
#         if not key in self.cache:
#             return -1
        
#         node = self.cache[key]
#         self.removeNode(node)
#         self.addToFront(node)

#         return node.val

#     def put(self, key: int, value: int) -> None:
#         if key in self.cache:
#             node = self.cache[key]
#             node.val = value
#             self.removeNode(node)
#             self.addToFront(node)
#         else:
#             if len(self.cache) == self.capacity:
#                 nodeToRemove = self.tail.prev
#                 self.removeNode(nodeToRemove)
#                 del self.cache[nodeToRemove.key]

#             node = Node(key, value)
#             self.cache[key] = node
#             self.addToFront(node)